administrator/components/com_media/controllers/api.php

```diff
@@ -0,0 +1,150 @@
+	 * Api endpoint for the media manager front end. The HTTP methods GET, PUT, POST and DELETE
+	 * are supported.
+	 *
+	 * The following query parameters are processed:
+	 * - path: The path of the resource, if not set then the default / is taken.
+	 *
+	 * Some examples with a more understandable rest url equivalent:
+	 * - GET a list of folders below the root:
+	 * 		index.php?option=com_media&task=api.files
+	 * 		/api/files
+	 * - GET a list of files and subfolders of a given folder:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop
+	 * 		/api/files/sampledata/fruitshop
+	 * - GET file information for a specific file:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop/test.jpg
+	 * 		/api/files/sampledata/fruitshop/test.jpg
+	 *
+	 * - POST a new file or folder into a specific folder:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop
+	 * 		/api/files/sampledata/fruitshop
+	 *
+	 * - PUT a media file:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop/test.jpg
+	 * 		/api/files/sampledata/fruitshop/test.jpg
+	 * - PUT process a media file:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop/test.jpg&action=process
+	 * 		/api/files/sampledata/fruitshop/test.jpg/process
+	 *
+	 * - DELETE an existing folder in a specific folder:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop/test
+	 * 		/api/files/sampledata/fruitshop/test
+	 * - DELETE an existing file in a specific folder:
+	 * 		index.php?option=com_media&task=api.files&path=/sampledata/fruitshop/test.jpg
+	 * 		/api/files/sampledata/fruitshop/test.jpg
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function files()
+	{
+		// @todo add ACL check
+
+		// Get the required variables
+		$path = $this->input->getPath('path', '/');
+
+		// Determine the method
+		$method = $this->input->getMethod() ? : 'GET';
+
+		try
+		{
+			// Gather the data accoring to the method
+			switch (strtolower($method))
+			{
+				case 'get':
+					$data = $this->adapter->getFiles($path);
+					break;
+				case 'delete':
+					$data = $this->adapter->delete($path);
+					break;
+				case 'post':
+				case 'put':
+				default:
+					throw new BadMethodCallException('Method not supported yet!');
+			}
+
+
+			// Return the data
+			$this->sendAndClose($data, 'ok', false);
+		}
+		catch (Exception $e)
+		{
+			$this->sendAndClose(null, 'Error: ' . $e->getMessage(), true);
+		}
+	}
+
+	/**
+	 * Echoes the given data as JSON in the following format:
+	 *
+	 * {"success":true,"message":"ok","messages":null,"data":[{"type":"dir","name":"banners","path":"//"}]}
+	 *
+	 * @param   mixed    $data     The data to send
+	 * @param   string   $message  The message
+	 * @param   boolean  $error    If it is an error response
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private function sendAndClose($data, $message, $error)
```

Comment:
General question, should we stay on protected for all intern methods? So we still have the ability to override them?